src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/OperationFormatter.cs

```diff
@@ -342,6 +343,22 @@ private void SerializeBodyContents(XmlDictionaryWriter writer, MessageVersion ve
             SerializeBody(writer, version, RequestAction, messageDescription, returnValue, parameters, isRequest);
         }
 
+        private async Task SerializeBodyContentsAsync(XmlDictionaryWriter writer, MessageVersion version, object[] parameters, object returnValue, bool isRequest)
+        {
+            MessageDescription messageDescription;
+            StreamFormatter streamFormatter;
+
+            SetupStreamAndMessageDescription(isRequest, out streamFormatter, out messageDescription);
+
+            if (streamFormatter != null)
+            {
+                await streamFormatter.SerializeAsync(writer, parameters, returnValue);
+                return;
+            }
+
+            SerializeBody(writer, version, RequestAction, messageDescription, returnValue, parameters, isRequest);
```

Comment:
I like @iamjasonp 's proposal.  I recommend checking in as-is and opening a new issue to make the appropriate virtual methods on OperationFormatter so the derived classes can do different things.